File: Data Structures & Algorithms/coin-change/submission-1.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0

        dp = [float('inf') for _ in range(amount + 1)]

        for ci in coins:
            if ci > amount:
                continue
            dp[ci] = 1

        for i in range(amount + 1):
            if (dp[i] == -1):
                continue

            for ci in coins:
                if i + ci > amount:
                    continue
                
                dp[i + ci] = min(dp[i + ci], 1 + dp[i])
        
        if dp[amount] == float('inf'):
            return -1

        return dp[amount]

File: Data Structures & Algorithms/jump-game-ii/submission-1.cpp
#include <algorithm>
#include <climits>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int jump(vector<int> &nums) {
    if (nums.size() == 1)
      return 0;

    int currIdx = 0;
    int jumps = 0;
    while (true) {
      if (nums[currIdx] + currIdx >= nums.size() - 1)
        return jumps + 1;

      int bestRangeIdx = -1;
      int currBestRange = INT_MIN;
      // for every possible jump
      for (int j = currIdx + 1;
           j <= min(currIdx + nums[currIdx], (int)nums.size()); ++j) {

        // record the one with max next range
        if (nums[j] + j > currBestRange) {
          currBestRange = nums[j] + j;
          bestRangeIdx = j;
        }
      }

      // cout << currIdx << " -> " << bestRangeIdx << endl;
      currIdx = bestRangeIdx;
      jumps++;
    }

    return jumps;
  }
};